pkg/controllers/disruption/tracker.go

```diff
@@ -0,0 +1,276 @@
+			// disabling to prevent panics later.
+			enabled = false
+		}
+	}
+	tracker := &Tracker{
+		// nolint:staticcheck
+		decisionCache:       decisionCache,
+		disruptionKeyMap:    make(map[string]string),
+		bucketsThresholdMap: bucketsThresholdMap,
+		enabled:             enabled,
+		cluster:             cluster,
+		clock:               clock,
+	}
+	return tracker
+}
+
+func (t *Tracker) GatherClusterState(ctx context.Context) (*ClusterStateSnapshot, error) {
+	deployments := appsv1.DeploymentList{}
+	err := t.client.List(ctx, &deployments)
+	if err != nil {
+		return nil, err
+	}
+
+	totalDesiredPodCount := 0
+	totalDesiredPodCount = lo.SumBy(deployments.Items, func(d appsv1.Deployment) int {
+		return int(*d.Spec.Replicas)
+	})
+
+	rs := lo.Map(deployments.Items, func(d appsv1.Deployment, _ int) corev1.ResourceList {
+		requests := lo.Map(d.Spec.Template.Spec.Containers, func(c corev1.Container, _ int) corev1.ResourceList {
+			return c.Resources.Requests
+		})
+
+		totalContainerResources := resources.Merge(requests...)
+
+		for i, r := range totalContainerResources {
+			r.Set(r.Value() * int64(*d.Spec.Replicas))
+			totalContainerResources[i] = r
+		}
+
+		return totalContainerResources
+	})
+
+	totalDesiredPodResources := resources.Merge(rs...)
+
+	return &ClusterStateSnapshot{
+		clusterCost:              t.clusterCost.GetClusterCost(),
+		podResources:             t.cluster.GetTotalPodResourceRequests(),
+		totalNodes:               len(t.cluster.DeepCopyNodes()),
+		totalDesiredPodCount:     totalDesiredPodCount,
+		totalDesiredPodResources: &totalDesiredPodResources,
+	}, nil
+}
+
+func (t *Tracker) AddCommand(ctx context.Context, cmd *Command) {
+	if !t.enabled {
+		return
+	}
+	s, err := t.GatherClusterState(ctx)
+	if err != nil {
+		log.FromContext(ctx).Error(err, "gathering cluster state", "cmd", cmd)
+		DecisionTrackerErrors.Inc(map[string]string{})
+	}
+	d := TrackedDecision{
+		clusterState: s,
+		startTime:    t.clock.Now(),
+		command:      cmd,
+		otherKeys:    lo.Map(cmd.Candidates, func(c *Candidate, _ int) string { return c.NodeClaim.Name }),
+	}
+	t.Lock()
+	defer t.Unlock()
+	for _, k := range d.otherKeys {
+		err := t.decisionCache.Add(k, d, 0)
+		if err != nil {
+			log.FromContext(ctx).Error(err, "failed to add decision to cache", "key", k)
+			DecisionTrackerErrors.Inc(map[string]string{})
+		}
+	}
+}
+
+func (t *Tracker) FinishCommand(ctx context.Context, nc *v1.NodeClaim) {
+	if !t.enabled {
+		return
+	}
+	s, err := t.GatherClusterState(ctx)
+	if err != nil {
+		log.FromContext(ctx).Error(err, "gathering cluster state", "nc", nc)
+		DecisionTrackerErrors.Inc(map[string]string{})
+	}
+	t.Lock()
+
+	d, exists := t.decisionCache.Get(nc.Name)
+	if !exists {
+		log.FromContext(ctx).Error(fmt.Errorf("decision not found in cache for NodeClaim"), "nc", nc)
+		DecisionTrackerErrors.Inc(map[string]string{})
+		t.Unlock()
+		return
+	}
+	decision, okay := d.(*TrackedDecision)
```

Comment:
nit: standard naming
```suggestion
	decision, ok := d.(*TrackedDecision)
```